### PR TITLE
SONAR-7794 - WS api/qualityprofiles/search return userUpdatedAt

### PR DESCRIPTION
**Pull request by teryk**
please review the commits regarding SONAR-7794. There are 4 commits

### CODE REVIEW

File: sonar-db/src/main/java/org/sonar/db/qualityprofile/QualityProfileDto.java [hunk 1]
```diff
@@ -123,6 +124,16 @@ public QualityProfileDto setLastUsed(@Nullable Long lastUsed) {
     return this;
   }
 
+  @CheckForNull
+  public Long getUserUpdatedAt() {
+    return userUpdatedAt;
+  }
+
+  public QualityProfileDto setUserUpdatedAt(@Nullable Long userUpdatedAt) {
```
Comment (julienlancelot) [inline sonar-db/src/main/java/org/sonar/db/qualityprofile/QualityProfileDto.java]: When this column can be null ?

File: sonar-db/src/main/java/org/sonar/db/version/v60/PopulateUserUpdatedAtOfRulesProfiles.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+import javax.annotation.CheckForNull;
+import org.sonar.db.Database;
+import org.sonar.db.version.BaseDataChange;
+import org.sonar.db.version.MassUpdate;
+import org.sonar.db.version.Select;
+import org.sonar.db.version.SqlStatement;
+
+public class PopulateUserUpdatedAtOfRulesProfiles extends BaseDataChange {
+
+  private static final String SQL_SELECT_PROFILES_NOT_UPDATED = "select kee from rules_profiles where user_updated_at is null";
```
Comment (julienlancelot) [inline sonar-db/src/main/java/org/sonar/db/version/v60/PopulateUserUpdatedAtOfRulesProfiles.java]: I prefer when SQL keywords are in uppercase

File: sonar-db/src/main/java/org/sonar/db/version/v60/PopulateUserUpdatedAtOfRulesProfiles.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    } catch (SQLException e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  private static void populateUserUpdatedAtColumn(Context context, Map<String, Long> userUpdatedAdByProfileKey) throws SQLException {
+    MassUpdate massUpdate = context.prepareMassUpdate();
+    massUpdate.select(SQL_SELECT_PROFILES_NOT_UPDATED);
+    massUpdate.update("update rules_profiles set user_updated_at=? where kee=?");
+    massUpdate.rowPluralName("rules_profiles");
```
Comment (julienlancelot) [inline sonar-db/src/main/java/org/sonar/db/version/v60/PopulateUserUpdatedAtOfRulesProfiles.java]: The name here is a human readable name that will be display in the logs of the server, I would prefer to have something like "quality profiles without user updated at date"

Comment (julienlancelot) [inline sonar-db/src/main/java/org/sonar/db/version/v60/PopulateUserUpdatedAtOfRulesProfiles.java]: or only "quality profiles"

File: sonar-db/src/main/java/org/sonar/db/qualityprofile/QualityProfileDto.java (hunk 1 above)
Comment (julienlancelot) [inline sonar-db/src/main/java/org/sonar/db/qualityprofile/QualityProfileDto.java]: Forgot this question. when no modification has been made by a user on a profile, this column will be null

**Comment by julienlancelot** [comment]
I'm wondering : we know when a user has updated a profile, but not who. Do you know if who has updated the profile could be useful one day ?

File: server/sonar-server/src/test/java/org/sonar/server/qualityprofile/RuleActivatorMediumTest.java [hunk 4]
```diff
@@ -1138,4 +1148,10 @@ private void verifyZeroActiveRules(String key) {
         .setQProfileKey(key)
         .setActivation(true))).isEmpty();
   }
+
+  private void assertProfileHasBeenUpdatedManually(String profileKey) {
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/qualityprofile/RuleActivatorMediumTest.java]: You should add a test where you check that profile is not manually updated when it was not updated by a user

File: server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/SearchActionTest.java [hunk 5]
```diff
@@ -128,21 +128,23 @@ public void search_map_dates() {
...
+    assertThat(parseDateTime(result.getProfiles(0).getUserUpdatedAt()).getTime()).isEqualTo(time);
   }
 
   @Test
   public void search_for_language() throws Exception {
     qualityProfileDb.insertQualityProfiles(QualityProfileDto.createFor("sonar-way-xoo1-12345").setLanguage(xoo1.getKey()).setName("Sonar way"));
 
     String result = ws.newRequest().setParam("language", xoo1.getKey()).execute().getInput();
-
+§
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/SearchActionTest.java]: Please remove this character

File: server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/SearchActionTest.java [hunk 6]
```diff
@@ -128,21 +128,23 @@ public void search_map_dates() {
...
-      .setLastUsed(time));
+      .setLastUsed(time)
+      .setUserUpdatedAt(time));
 
     SearchWsResponse result = call(ws.newRequest());
 
     assertThat(result.getProfilesCount()).isEqualTo(1);
     assertThat(result.getProfiles(0).getRulesUpdatedAt()).isEqualTo("2016-12-21T19:10:03+0100");
     assertThat(parseDateTime(result.getProfiles(0).getLastUsed()).getTime()).isEqualTo(time);
+    assertThat(parseDateTime(result.getProfiles(0).getUserUpdatedAt()).getTime()).isEqualTo(time);
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/SearchActionTest.java]: You should add a test (or a check in another test) that userUpdatedAt is not returned when the profile was not manually updated

**Comment by julienlancelot** [comment]
LGTM

File: sonar-db/src/main/java/org/sonar/db/version/v60/PopulateUserUpdatedAtOfRulesProfiles.java (hunk 3 above)
Comment (teryk) [inline sonar-db/src/main/java/org/sonar/db/version/v60/PopulateUserUpdatedAtOfRulesProfiles.java]: ok

**Comment by teryk** [comment]
- maybe we'll need the author of the last modification
- maybe we'll need the project where the quality profile was used for the last time

then, we'll need to change the design, for example by creating a specific db table

File: server/sonar-server/src/test/java/org/sonar/server/qualityprofile/RuleActivatorMediumTest.java (hunk 4 above)
Comment (teryk) [inline server/sonar-server/src/test/java/org/sonar/server/qualityprofile/RuleActivatorMediumTest.java]: ok

File: server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/SearchActionTest.java (hunk 5 above)
Comment (teryk) [inline server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/SearchActionTest.java]: ok

File: server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/SearchActionTest.java (hunk 6 above)
Comment (teryk) [inline server/sonar-server/src/test/java/org/sonar/server/qualityprofile/ws/SearchActionTest.java]: here's it's a search. I check it maps the data correctly. We don't care about the profile being manually updated or not in this context

File: server/sonar-server/src/test/java/org/sonar/server/activity/ws/ActivitiesWsMediumTest.java [hunk 7]
```diff
@@ -70,6 +70,7 @@ public void search() throws Exception {
     activity.setMessage("THE_MSG");
     activity.setData("foo", "bar");
     activity.setData("profileKey", "PROFILE_KEY");
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/activity/ws/ActivitiesWsMediumTest.java]: This line should be removed

File: server/sonar-server/src/test/java/org/sonar/server/activity/ActivityServiceTest.java [hunk 8]
```diff
@@ -72,6 +72,7 @@ public void insert_and_index() {
     activity.setMessage("THE_MSG");
     activity.setData("foo", "bar");
     activity.setData("profileKey", "PROFILE_KEY");
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/activity/ActivityServiceTest.java]: This line should be removed

Comment (teryk) [inline server/sonar-server/src/test/java/org/sonar/server/activity/ActivityServiceTest.java]: it is removed in a later commit

File: server/sonar-server/src/test/java/org/sonar/server/activity/ws/ActivitiesWsMediumTest.java (hunk 7 above)
Comment (teryk) [inline server/sonar-server/src/test/java/org/sonar/server/activity/ws/ActivitiesWsMediumTest.java]: ok

File: server/sonar-server/src/test/java/org/sonar/server/activity/ActivityServiceTest.java [hunk 9]
```diff
@@ -71,15 +71,21 @@ public void insert_and_index() {
...
+      "  log_action as \"action\", " +
+      "  log_message as \"msg\", " +
+      "  data_field as \"data\", " +
+      "  profile_key as \"profileKey\" " +
+      "from activities");
     assertThat(dbMap).containsEntry("type", "ANALYSIS_REPORT");
     assertThat(dbMap).containsEntry("action", "THE_ACTION");
     assertThat(dbMap).containsEntry("msg", "THE_MSG");
-    assertThat(dbMap.get("data")).isEqualTo("foo=bar;profileKey=PROFILE_KEY");
+    assertThat(dbMap).containsEntry("profileKey", "PROFILE_KEY");
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/activity/ActivityServiceTest.java]: dbMap should not contains profileKey

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/RuleActivator.java [hunk 10]
```diff
@@ -63,17 +64,19 @@
   private final RuleIndex ruleIndex;
   private final ActiveRuleIndexer activeRuleIndexer;
   private final ActivityService activityService;
+  private final UserSession userSession;
 
   public RuleActivator(System2 system2, DbClient db, RuleIndex ruleIndex,
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/RuleActivator.java]: MAJOR Constructor has 8 parameters, which is greater than 7 authorized. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS00107)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: server/sonar-server/src/test/java/org/sonar/server/qualityprofile/RuleActivatorMediumTest.java [hunk 11]
```diff
@@ -1059,6 +1059,9 @@ public void bulk_change_severity() {
     verifyHasActiveRuleInDbAndIndex(ActiveRuleKey.of(XOO_P2_KEY, XOO_X2), BLOCKER, INHERITED, Collections.<String, String>emptyMap());
     verifyHasActiveRuleInDbAndIndex(ActiveRuleKey.of(XOO_P3_KEY, XOO_X1), BLOCKER, INHERITED, ImmutableMap.of("max", "7"));
     verifyHasActiveRuleInDbAndIndex(ActiveRuleKey.of(XOO_P3_KEY, XOO_X2), BLOCKER, INHERITED, Collections.<String, String>emptyMap());
+    assertProfileHasBeenUpdatedAutomatically(XOO_P1_KEY);
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/qualityprofile/RuleActivatorMediumTest.java]: Why in bulk change the profile is not considered as having been updated manually ?

Comment (teryk) [inline server/sonar-server/src/test/java/org/sonar/server/qualityprofile/RuleActivatorMediumTest.java]: because userSession hasn't been set

**Comment by julienlancelot** [comment]
LGTM

**Pull request closed without merging**